eks_infra/upgrade_cluster.sh

```diff
@@ -0,0 +1,173 @@
+}
+
+# Function to create static and dynamic nodegroups in EKS cluster
+function create_nodegroups(){
+    #static nodegroup
+    eksctl create nodegroup \
+    --name static-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type m5.large \
+    --nodes 1 \
+    --node-labels "static=true" \
+    --tags "k8s.io/cluster-autoscaler/node-template/label/static=true" \
+    --asg-access \
+    --ssh-access \
+    --ssh-public-key "${3}"
+
+    #dynamic gpu nodegroup
+    eksctl create nodegroup \
+    --name gpu-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type p3.16xlarge \
+    --nodes-min 0 \
+    --nodes-max 100 \
+    --node-volume-size 80 \
+    --node-labels "test_type=gpu" \
+    --tags "k8s.io/cluster-autoscaler/node-template/label/test_type=gpu" \
+    --asg-access \
+    --ssh-access \
+    --ssh-public-key "${3}"
+
+    # dynamic inf nodegroup
+  
+    eksctl create nodegroup \
+    --name inf-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type inf1.xlarge \
+    --nodes-min 0 \
+    --nodes-max 100 \
+    --node-volume-size 500 \
+    --node-ami "ami-092059396c7e51f52" \
+    --node-labels "test_type=inf" \
+    --tags "k8s.io/cluster-autoscaler/node-template/label/test_type=inf" \
+    --asg-access \
+    --ssh-access \
+    --ssh-public-key "${3}"
+  
+}
+
+# Function to delete all nodegroups in EKS cluster
+function delete_nodegroups(){
+
+    LIST_NODE_GROUPS=$(eksctl get nodegroup --cluster ${1} -o json | jq -r '.[].Name')
+
+    if [ -n "${LIST_NODE_GROUPS}" ]; then
+    
+      for NODEGROUP in $LIST_NODE_GROUPS; do
+        eksctl delete nodegroup \
+        --name $NODEGROUP \
+        --cluster ${1} \
+        --region ${2}
+      done
+    else
+      echo "No Nodegroups present in the EKS cluster ${1}"
+    fi
+}
+
+# Function to upgrade nodegroups
+function upgrade_nodegroups(){
+    delete_nodegroups ${1} ${3}
+    create_nodegroups ${1} ${2} ${4}
+}
+
+#Function ton upgrade core k8s components
+function update_eksctl_utils(){
+    eksctl utils update-kube-proxy \
+    --cluster ${1} \
+    --region ${2} \
+    --approve
+
+    eksctl utils update-aws-node \
+    --cluster ${1} \
+    --region ${2} \
+    --approve
+
+    eksctl utils update-coredns \
+    --cluster ${1} \
+    --region ${2} \
+    --approve
+}
+
+if [ $# -ne 3 ]; then
+    echo "${0}: usage: ./upgrade_cluster.sh eks_cluster_name eks_version cluster_autoscalar_image_version"
+    exit 1
+fi
+
+if [ -z "${AWS_REGION}" ]; then
+  echo "AWS region not configured"
+  exit 1
+fi
+
+if [ -z "${EKS_CLUSTER_MANAGER_ROLE}" ]; then
+  echo "EKS cluster management role not set"
+  exit 1
+fi
+
+CLUSTER=${1}
+EKS_VERSION=${2}
+CLUSTER_AUTOSCALAR_IMAGE_VERSION=${3}
+REGION=${AWS_REGION}
+EKS_ROLE=${EKS_CLUSTER_MANAGER_ROLE}
+
+if [ -z "${EC2_KEY_PAIR_NAME}" ]; then
+  KEY_NAME=${CLUSTER}-KeyPair
+  echo "No EC2 key pair name configured. Creating KeyPair ${KEY_NAME}"
+  create_ec2_key_pair ${KEY_NAME}
+  EC2_KEY_PAIR_NAME=${KEY_NAME}
```

Comment:
Is there going to be a case where there already exists a cluster for which a keypair doesn't exist?

If whoever executes the `upgrade` forgets to configure the custom `EC2_KEY_PAIR_NAME` they previously used to create their cluster, they could inadvertently change the SSH Keypair used for their cluster without even knowing it after the change.

Reply:
Keypairs are dedicated to nodegroups i.e. the launch template used by autoscaling group to spin up EC2 instance. It is not related to an EKS cluster. Hence, we can have nodegroups with dedicated keypairs assigned to each of them. 

In this case, during an upgrade, if a user do not have a keypair used previously, a new keypair can be created and assigned to nodegroup.